templates/saveImage.php

```diff
@@ -51,10 +51,20 @@ function validateAndSaveImage($file, $maxSize, $allowedWidth, $allowedHeight, $d
     $actualWidth = imagesx($image);
     $actualHeight = imagesy($image);
 
-    $size = min($actualWidth, $actualHeight);
+    //Determine the new size for the image
+    $xSize = min($actualWidth, $actualHeight);
+    $ySize = $xSize;
 
-    $resizedImage = imagecreate($allowedWidth, $allowedHeight);
-    imagecopyresized($resizedImage, $image, 0, 0, $actualWidth / 2 - $size / 2, $actualHeight / 2 - $size / 2, $allowedWidth, $allowedHeight, $size, $size);
+    //Determine if we want to keep origina dimensions
+    if ($allowedWidth <= 0)
+        $xSize = $actualWidth;
+
+    if ($allowedHeight <= 0)
+        $ySize = $actualHeight;
+
+    $image = imagecrop($image, ['x' => $actualWidth / 2 - $xSize / 2, 'y' => $actualHeight / 2 - $ySize / 2, 'width' => $xSize, 'height' => $ySize]);
+
+    imagecopyresized($image, $image, 0, 0, 0, 0, $xSize, $ySize, $actualWidth, $actualHeight);
```

Comment:
```suggestion
  // check image size
  $actualWidth = imagesx($image);
  $actualHeight = imagesy($image);
  
  //Get original dimensions
  $xSize = $actualWidth;
  $ySize = $actualHeight;
  
  //Determine the new size for the image
  if ($allowedWidth && $allowedHeight){
      //Find ratio between aspect ratios
      $aspectRatio = $actualHeight / $actualWidth;
      $desiredRatio = $allowedHeight / $allowedWidth;
  
      $change = $aspectRatio / $desiredRatio;
      //Adjust dimensions to meet new aspect ratio
      if ($change > 1){
          $ySize /= $change;
      }else{
          $xSize *= $change;
      }
  }
  
  //Determine if we want to keep original dimensions
  if ($allowedWidth <= 0) $allowedWidth = $actualWidth;
  if ($allowedHeight <= 0) $allowedHeight = $actualHeight;
      
  $resizedImage = imagecreate($allowedWidth, $allowedHeight);

  imagecopyresized($resizedImage, $image, 0, 0, $actualWidth / 2 - $xSize / 2, $actualHeight / 2 - $ySize / 2, $allowedWidth, $allowedHeight, $xSize, $ySize);
```
Your changes here solve the problem for receipt uploads but mess up profile picture uploads. This algorithm 'should' work for both I think. Also the image copy resize currently does nothing (an error from when it was originally written) so I added a new resizedImage variable that should contain the actually resized image, The two references to $image after this need to be swapped to $resizedImage